File: 206. Reverse Linked List/python/reverse_linked_list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseListS(self, head: Optional[ListNode]) -> Optional[ListNode]:
        ch = head
        resNodes = None
        while ch is not None:
            tmp_ch = ch.next
            ch.next = resNodes
            resNodes = ch
            ch = tmp_ch
        return resNodes
    
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        ch = head
        resNodes = None
        while ch is not None:
            resNodes = ListNode(val=ch.val, next=resNodes)
            ch = ch.next
        return resNodes

File: 206. Reverse Linked List/go/reverse_linked_list.go
package reverseLinkedList

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseListOne(head *ListNode) *ListNode {
	ch := head
	var resNodes *ListNode
	for ch != nil {
		tmpCh := ch.Next
		ch.Next = resNodes
		resNodes = ch
		ch = tmpCh
	}
	return resNodes
}

func reverseList(head *ListNode) *ListNode {
	ch := head
	var resNodes *ListNode
	for ch != nil {
		preRes := ListNode{Val: ch.Val, Next: resNodes}
		resNodes = &preRes
		ch = ch.Next
	}
	return resNodes
}
